File: ts/packages/over-redis/src/backoff.ts
export interface BackoffOptions {
  initialMs: number;
  maxMs: number;
  factor?: number;
  jitter?: boolean;
}

export function expBackoff(opts: BackoffOptions): () => number {
  const factor = opts.factor ?? 2;
  let current = opts.initialMs;
  return () => {
    const value = current;
    current = Math.min(opts.maxMs, current * factor);
    if (opts.jitter ?? true) return Math.floor(value * (0.5 + Math.random() * 0.5));
    return value;
  };
}

export function resetable(opts: BackoffOptions): { next: () => number; reset: () => void } {
  let next = expBackoff(opts);
  return {
    next: () => next(),
    reset: () => { next = expBackoff(opts); },
  };
}

File: py/packages/over-redis/src/clamator_over_redis/backoff.py
import random
from dataclasses import dataclass


@dataclass
class ExpBackoff:
    initial_ms: int
    max_ms: int
    factor: float = 2.0
    jitter: bool = True
    _current_ms: int | None = None

    def next_delay_ms(self) -> int:
        if self._current_ms is None:
            self._current_ms = self.initial_ms
        value = self._current_ms
        self._current_ms = min(self.max_ms, int(self._current_ms * self.factor))
        if self.jitter:
            return int(value * (0.5 + random.random() * 0.5))
        return value

    def reset(self) -> None:
        self._current_ms = None
